tools/SwiftSyntax/SyntaxRewriter.swift.gyb

```diff
@@ -45,13 +45,29 @@ open class SyntaxRewriter {
   ///   - node: the node we are about to visit.
   open func visitPre(_ node: Syntax) {}
 
+  /// Override point to choose custom visitation dispatch instead of the
+  /// specialized `visit(_:)` methods. Use this instead of those methods if
+  /// you intend to dynamically dispatch rewriting behavior.
+  /// - note: If this method returns a non-nil result, the specialized
+  ///         `visit(_:)` methods will not be called for this node.
+  open func visitAny(_ node: Syntax) -> Syntax? {
+    return nil
+  }
+
   /// The function called after visting the node and its descendents.
   ///   - node: the node we just finished visiting.
   open func visitPost(_ node: Syntax) {}
 
   public func visit(_ node: Syntax) -> Syntax {
     visitPre(node)
     defer { visitPost(node) }
+
+    // If the global visitor returned non-nil, visit the rewritten node's
+    // children and skip specialized dispatch.
+    if let newNode = visitAny(node) {
+      return visitChildren(newNode)
```

Comment:
hmm, sorry about being a little nit-picky here. Whether `visitingChildren` should be up to the overriding function, right? like other specialized visit functions.

Reply:
Oh, yes you're right. Thanks for catching that!

Reply:
The more I think about it, though, the more I wonder if we _should_ always call visitChildren after the specialized visitor has been called.

For example, if I'm writing a transform that, say, lays out function call arguments a certain way, I shouldn't have to manually iterate over all subexpressions and visit them, no?

Reply:
yeah, i think it is a valid assumption that in most cases users will visit a node's children even though the node itself has been updated. @rintaro what do you think?

Reply:
Perhaps, we should separate `Visitor` and `Walker` like `ASTVisitor` / `ASTWalker`.